### ISSUE TITLE
Security issue with .bat file

### ISSUE THREAD

**Issue by petems** [WinRb/vagrant-windows#160]
I think this has been answered before (from the comment on https://github.com/mitchellh/vagrant/issues/2395) but I can't find the solution.

Basically I'm having an issue with some puppet code when I'm provisioning a Vagrant instance:

```
Error: /Stage[main]/Qix::Config/File[c:/foo/bar/baz.bat]: Could not evaluate: Failed to get security information:  Incorrect function. Could not retrieve file metadata for puppet:///modules/qix/baz.bat: Failed to get security information:  Incorrect function.
```

``` puppet
file { 'c:/foo/bar/baz.bat':
    ensure => present,
    source => 'puppet:///modules/qix/baz.bat',
    group  => 'SYSTEM',
    owner  => 'Administrators',
    mode   => '0777',
}
```

Any help appreciated :+1:

**Comment by sneal** [comment]
I'm not really very familiar with the Puppet source code, so I can't really say with any authority what's going on. 

My suspicion is that Puppet is trying to read the file permissions of the batch file which lives on the symlinked share that vagrant-windows creates to provide your scripts from your host to the guest. It seems odd it would need to do that, but completely possible.

If you attempt to look at any of the NTFS permissions for any of the vagrant-windows symlinked shares you'll notice there's absolutely no permissions and all the files show up with lock icons next to them in Windows Explorer.

Try adding a source_permissions attribute or removing the mode and see what happens.

```
file { 'c:/foo/bar/baz.bat':
    ensure => present,
    source => 'puppet:///modules/qix/baz.bat',
    group  => 'SYSTEM',
    owner  => 'Administrators',
    mode   => '0777',
    source_permissions => ignore,
}
```

**Comment by petems** [comment]
Cool, I ended up just changing it to a template and sourcing from that because I needed a quick win, but I'll try the source_permissions trick later! :+1:

**Comment by petems** [comment]
Ok, since this actually a puppet issue not a vagrant-windows issue, I'll close :+1:

**Comment by sneal** [comment]
source_permissions doesn't help things, although I would think Puppet should skip reading the source permissions if the target isn't going to use them.

**Comment by ferventcoder** [comment]
We are looking into this while researching [PUP-1389](https://tickets.puppetlabs.com/browse/PUP-1389)

**Comment by joshcooper** [comment]
I think the problem here is that puppet's logic for detecting if the source volume [supports NTFS ACLs](https://github.com/puppetlabs/puppet/blob/3.2.4/lib/puppet/util/windows/security.rb#L153) may not take into account that the target of the link may be on a different volume. Also, puppet does not pass the original [exception backtrace](https://github.com/puppetlabs/puppet/blob/3.2.4/lib/puppet/type/file/source.rb#L163) through, so it's hard to follow what's going on. We'll get that fixed.

**Comment by ferventcoder** [comment]
I'm seeing something similar with puppet 3.4.2 - 

```
Error: /Stage[main]/Main/File[c:/Bob/jenkins-slave/jenkins-slave.exe]: Could not evaluate: 
Invalid DACL:  The access control list (ACL) structure is invalid. Could not retrieve file 
metadata for puppet:///modules/jenkins/jenkins-slave.exe: 
Invalid DACL:  The access control list (ACL) structure is invalid.
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/util/errors.rb:97:in `fail'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file/source.rb:178:in `rescue in block in metadata'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file/source.rb:171:in `block in metadata'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file/source.rb:170:in `each'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file/source.rb:170:in `metadata'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file/source.rb:111:in `copy_source_values'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type/file.rb:633:in `retrieve'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/type.rb:1063:in `retrieve_resource'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/transaction/resource_harness.rb:223:in `from_resource'
C:/Program Files (x86)/Puppet Labs/Puppet/puppet/lib/puppet/transaction/resource_harness.rb:17:in `evaluate'

```

**Comment by Iristyle** [comment]
After doing some detective work, I just sent in the PR for the Puppet patch as https://github.com/puppetlabs/puppet/pull/2299
